auraed/src/observe/observed_event_stream.rs

```diff
@@ -0,0 +1,125 @@
+/* -------------------------------------------------------------------------- *\
+ *        Apache 2.0 License Copyright © 2022-2023 The Aurae Authors          *
+ *                                                                            *
+ *                +--------------------------------------------+              *
+ *                |   █████╗ ██╗   ██╗██████╗  █████╗ ███████╗ |              *
+ *                |  ██╔══██╗██║   ██║██╔══██╗██╔══██╗██╔════╝ |              *
+ *                |  ███████║██║   ██║██████╔╝███████║█████╗   |              *
+ *                |  ██╔══██║██║   ██║██╔══██╗██╔══██║██╔══╝   |              *
+ *                |  ██║  ██║╚██████╔╝██║  ██║██║  ██║███████╗ |              *
+ *                |  ╚═╝  ╚═╝ ╚═════╝ ╚═╝  ╚═╝╚═╝  ╚═╝╚══════╝ |              *
+ *                +--------------------------------------------+              *
+ *                                                                            *
+ *                         Distributed Systems Runtime                        *
+ *                                                                            *
+ * -------------------------------------------------------------------------- *
+ *                                                                            *
+ *   Licensed under the Apache License, Version 2.0 (the "License");          *
+ *   you may not use this file except in compliance with the License.         *
+ *   You may obtain a copy of the License at                                  *
+ *                                                                            *
+ *       http://www.apache.org/licenses/LICENSE-2.0                           *
+ *                                                                            *
+ *   Unless required by applicable law or agreed to in writing, software      *
+ *   distributed under the License is distributed on an "AS IS" BASIS,        *
+ *   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. *
+ *   See the License for the specific language governing permissions and      *
+ *   limitations under the License.                                           *
+ *                                                                            *
+\* -------------------------------------------------------------------------- */
+use super::{cgroup_cache::CgroupCache, proc_cache::ProcCache};
+use crate::ebpf::tracepoint::PerfEventBroadcast;
+use aurae_ebpf_shared::{HasCgroup, HasHostPid};
+use proto::observe::WorkloadType;
+use std::{ffi::OsString, sync::Arc};
+use tokio::sync::{
+    mpsc::{self, Receiver},
+    Mutex,
+};
+use tonic::Status;
+
+const CGROUPFS_ROOT: &str = "/sys/fs/cgroup";
+
+/// Wrapper around `PerfEventBroadvast<T>` that allows for filtering by
+/// Aurae workloads and optionally maps host PIDs to namespace PIDs.
+pub struct ObservedEventStream<'a, T> {
```

Comment:
I did a bit of research early on as I also preferred more functional style stream transformations instead of the imperative code that's in there right now (`while recv() { if pred { tx.send() } }`) but for some reason (likely the borrow checker :) ) I ended up with this. FWIW the `ObservedEventStream` itself is a very domain-specific type as it owns those caches and has a notion of filtering by workload and pid mapping so I'm not sure if we'll get away by replacing this with a tokio stream type without introducing a lot of extra code in the `ObserveService`. Maybe something to look into later?